csharp/ql/src/Telemetry/ExternalApiSpecific.qll

```diff
@@ -0,0 +1,80 @@
+/** Provides classes and predicates related to handling APIs from external libraries. */
+
+import csharp
+import semmle.code.csharp.dataflow.DataFlow
+import semmle.code.csharp.dataflow.ExternalFlow
+import semmle.code.csharp.dataflow.FlowSummary
+import semmle.code.csharp.dataflow.TaintTracking
+import semmle.code.csharp.security.dataflow.flowsources.Remote
+private import semmle.code.csharp.dataflow.internal.DataFlowPrivate
+private import semmle.code.csharp.dataflow.internal.TaintTrackingPrivate
+
+/**
+ * An external API from either the C# Standard Library or a 3rd party library.
+ */
+class ExternalApiSpecific extends Callable {
+  ExternalApiSpecific() { this.fromLibrary() }
+
+  /**
+   * Gets the unbound type, name and parameter types of this API.
+   */
+  private string getSignature() {
+    result =
+      this.getDeclaringType().getUnboundDeclaration() + "." + this.getName() + "(" +
+        this.parameterTypesToString() + ")"
+  }
+
+  /**
+   * Gets the namespace of this API.
+   */
+  private string getNamespace() { result = this.getDeclaringType().getNamespace().toString() }
+
+  /**
+   * Gets the assembly file name containing this API.
+   */
+  private string getAssembly() { result = this.getFile().getBaseName() }
+
+  /**
+   * Gets the assembly file name and namespace of this API.
+   */
+  string getInfoPrefix() { result = this.getAssembly() + "#" + this.getNamespace() }
+
+  /**
+   * Gets the assembly file name, namespace and signature of this API.
+   */
+  string getInfo() { result = this.getInfoPrefix() + "#" + this.getSignature() }
+
+  /** Gets a node that is an input to a call to this API. */
+  DataFlow::Node getAnInput() {
```

Comment:
The current implementation will not include `this` arguments. I think something like this should work


```ql
private import semmle.code.csharp.dataflow.internal.DataFlowDispatch
  DataFlow::ArgumentNode getAnInput() {
    exists(NonDelegateDataFlowCall call |
      result.getCall() = call and
      this = [call.getDispatchCall().getADynamicTarget(), call.getDispatchCall().getAStaticTarget()).getUnboundDeclaration()
    )
  }
```

Reply:
Could we do a knowledge sharing session on some of this? I have previously stared at IPA type TDataFlowCall and not understood, what was going on. It is not clear to me, why a solution (like the one outlined above) will work.

Reply:
Sure thing!